core/src/main/java/org/elasticsearch/common/cache/Cache.java

```diff
@@ -300,11 +323,8 @@ public V computeIfAbsent(K key, CacheLoader<K, V> loader) throws ExecutionExcept
         V value = get(key, now);
         if (value == null) {
             CacheSegment<K, V> segment = getCacheSegment(key);
-            // we synchronize against the segment lock; this is to avoid a scenario where another thread is inserting
-            // a value for the same key via put which would not be observed on this thread without a mechanism
-            // synchronizing the two threads; it is possible that the segment lock will be too expensive here (it blocks
-            // readers too!) so consider this as a possible place to optimize should contention be observed
-            try (ReleasableLock ignored = segment.writeLock.acquire()) {
+            K interned = segment.internKey(key);
+            synchronized (interned) {
```

Comment:
Yes, but I don't think that we want an internal lock per key. Back to the drawing board. :)